hibernate/hibernate-search
# HSEARCH-3335 Search 6 groundwork - Restore the distribution module

https://hibernate.atlassian.net//browse/HSEARCH-3335

I tried to improve the configuration while I was at it, so that we don't have to explicitly list the dependencies anymore: any new dependency will be added automatically. The only updates we will have to perform manually are:

* When we add a new module
* When we add a provided dependency
* When we change the inter-module dependencies (though if we forget to update in this case, the worst thing that can happen is that some JAR will appear twice)

To test it quickly, run this:

```
mvn clean install -DskipTests -Pdist -pl distribution -am -Dmaven.javadoc.skip=true
```

The distribution bundles will be in `distribution/target/dist/`

I'm not sure if we should merge the last commit, any opinion on that? I'll copy the commit message with the details:

> Not sure it's really an improvement, since the structure is
> significantly more complex, and I don't think consumers of this bundle
> will care much about which artifact comes from the Hibernate team and
> which one comes from an external source.
> 
> But at least that's closer to what we used to offer...
> 
> After this patch, the structure is as follows:
>
> ```
> $ tree hibernate-search-6.0.0-SNAPSHOT/dist
> hibernate-search-6.0.0-SNAPSHOT/dist
> ├── backend
> │   ├── elasticsearch
> │   │   ├── hibernate-search-backend-elasticsearch-6.0.0-SNAPSHOT.jar
> │   │   └── lib
> │   │       └── required
> │   │           ├── commons-codec-1.10.jar
> │   │           ├── commons-logging-1.1.3.jar
> │   │           ├── elasticsearch-rest-client-6.3.1.jar
> │   │           ├── elasticsearch-rest-client-sniffer-6.3.1.jar
> │   │           ├── gson-2.8.5.jar
> │   │           ├── httpasyncclient-4.1.2.jar
> │   │           ├── httpclient-4.5.2.jar
> │   │           ├── httpcore-4.4.5.jar
> │   │           ├── httpcore-nio-4.4.5.jar
> │   │           └── jackson-core-2.8.10.jar
> │   └── lucene
> │       ├── hibernate-search-backend-lucene-6.0.0-SNAPSHOT.jar
> │       └── lib
> │           └── required
> │               ├── lucene-analyzers-common-7.4.0.jar
> │               ├── lucene-core-7.4.0.jar
> │               ├── lucene-join-7.4.0.jar
> │               └── lucene-sandbox-7.4.0.jar
> ├── engine
> │   ├── hibernate-search-engine-6.0.0-SNAPSHOT.jar
> │   ├── hibernate-search-util-internal-common-6.0.0-SNAPSHOT.jar
> │   └── lib
> │       └── required
> │           └── jboss-logging-3.3.2.Final.jar
> ├── mapper
> │   └── orm
> │       ├── hibernate-search-mapper-pojo-6.0.0-SNAPSHOT.jar
> │       ├── hibernate-search-orm-6.0.0-SNAPSHOT.jar
> │       └── lib
> │           └── required
> │               ├── antlr-2.7.7.jar
> │               ├── byte-buddy-1.8.17.jar
> │               ├── classmate-1.3.4.jar
> │               ├── dom4j-1.6.1.jar
> │               ├── hibernate-commons-annotations-5.0.4.Final.jar
> │               ├── hibernate-core-5.3.6.Final.jar
> │               ├── jandex-2.0.5.Final.jar
> │               ├── javassist-3.23.1-GA.jar
> │               ├── javax.activation-api-1.2.0.jar
> │               └── xml-apis-1.0.b2.jar
> └── provided
>     ├── cdi-api-2.0.jar
>     ├── javax.batch-api-1.0.jar
>     ├── javax.inject-1.jar
>     ├── javax.persistence-api-2.2.jar
>     ├── jboss-annotations-api_1.2_spec-1.0.2.Final.jar
>     └── jboss-transaction-api_1.2_spec-1.1.1.Final.jar
> ```
> 
> For reference, before this patch, the structure was something like this:
> 
> ```
> $ tree hibernate-search-6.0.0-SNAPSHOT/dist
> hibernate-search-6.0.0-SNAPSHOT/dist
> ├── backend
> │   ├── elasticsearch
> │   │   ├── commons-codec-1.10.jar
> │   │   ├── commons-logging-1.1.3.jar
> │   │   ├── elasticsearch-rest-client-6.3.1.jar
> │   │   ├── elasticsearch-rest-client-sniffer-6.3.1.jar
> │   │   ├── gson-2.8.5.jar
> │   │   ├── hibernate-search-backend-elasticsearch-6.0.0-SNAPSHOT.jar
> │   │   ├── httpasyncclient-4.1.2.jar
> │   │   ├── httpclient-4.5.2.jar
> │   │   ├── httpcore-4.4.5.jar
> │   │   ├── httpcore-nio-4.4.5.jar
> │   │   └── jackson-core-2.8.10.jar
> │   └── lucene
> │       ├── hibernate-search-backend-lucene-6.0.0-SNAPSHOT.jar
> │       ├── lucene-analyzers-common-7.4.0.jar
> │       ├── lucene-core-7.4.0.jar
> │       ├── lucene-join-7.4.0.jar
> │       └── lucene-sandbox-7.4.0.jar
> ├── engine
> │   ├── hibernate-search-engine-6.0.0-SNAPSHOT.jar
> │   ├── hibernate-search-util-internal-common-6.0.0-SNAPSHOT.jar
> │   └── jboss-logging-3.3.2.Final.jar
> ├── mapper
> │   └── orm
> │       ├── antlr-2.7.7.jar
> │       ├── byte-buddy-1.8.17.jar
> │       ├── classmate-1.3.4.jar
> │       ├── dom4j-1.6.1.jar
> │       ├── hibernate-commons-annotations-5.0.4.Final.jar
> │       ├── hibernate-core-5.3.6.Final.jar
> │       ├── hibernate-search-mapper-pojo-6.0.0-SNAPSHOT.jar
> │       ├── hibernate-search-orm-6.0.0-SNAPSHOT.jar
> │       ├── jandex-2.0.5.Final.jar
> │       ├── javassist-3.23.1-GA.jar
> │       ├── javax.activation-api-1.2.0.jar
> │       └── xml-apis-1.0.b2.jar
> └── provided
>     ├── cdi-api-2.0.jar
>     ├── javax.batch-api-1.0.jar
>     ├── javax.inject-1.jar
>     ├── javax.persistence-api-2.2.jar
>     ├── jboss-annotations-api_1.2_spec-1.0.2.Final.jar
>     └── jboss-transaction-api_1.2_spec-1.1.1.Final.jar
> ```
